### ISSUE TITLE
Bug: on_finish (trial) and on_finish_trial callbacks and data.addDataToLastTrial

### ISSUE THREAD

**Issue by wolfgangwalther** [jodeleeuw/jsPsych#300]
Using the following code:

``` javascript
var trial = {
  type = 'anything',
  on_finish: function(data) {
    data.variable1 = 'value1';
    jsPsych.data.addDataToLastTrial({variable2: 'value2'});
  };

jsPsych.init({
  timeline: [...],
  on_finish_trial: function(data) {
    console.log(data.variable1); // returns undefined, variable1 is not set at all
    console.log(data.variable2); // returns undefined, but variable2 is actually set
  }
});
```

The two callbacks are called in this order: `on_finish` (trial) and then `on_finish_trial`. Both have the data parameter passed to them as a deep copy (see [finishTrial](https://github.com/jodeleeuw/jsPsych/blob/master/jspsych.js#L176) and [module.getData](https://github.com/jodeleeuw/jsPsych/blob/master/jspsych.js#L881)). Therefore, setting `data.variable1` in `on_finish` does not change the underlying data object. This seems to be expected behaviour (although this was different in the past ~ jsPsych 5.x, where the trial's data could be modified in the `on_finish` callbacks).

Because the `trial_data` object is not loaded again between the two callbacks, the `data` parameter in `on_finish_trial` doesn't contain the new data, that has been updated with `data.addDataToLastTrial(...)`.

As a side note: I'm using the `on_trial_finish` callback to send the trial data to the server after each trial. Currently, this means that I can't update the data in the `on_finish` callback and then have the correct values send to the server without some workarounds. There are different ways to do it, but I solved for now, by passing the data object by reference instead of as a deep copy in the core. This doesn't seem ideal, because although I don't know for sure, I guess there was good reason for this to be changed from the earlier behaviour in ~ jsPsych 5.x.
Conceptually it would make more sense to use the `on_data_update` callback to send my trial data to the server. But since both of the other callbacks are executed after the `data.write` call, I wouldn't be able to clean up my data anymore before sending.

**Long story short:** I suggest to change `finishTrial`, so that the `on_finish` callback is called before `data.write` and given the `data` object by reference, so that this trial-level callback can modify the data before it is saved.
The `on_trial_finish` can then be called after `data.write` and be passed the `data` parameter as a copy. The data would already have been written, so it makes sense, that it can't be modified here anymore.

There are other solutions as well, like going back to passing by reference in general in the `data` module functions or staying with the deep copy approach, but reading the data again between the two callbacks.

Any thoughts on this? What would be the smartest way to handle this?

**Comment by jodeleeuw** [comment]
> I suggest to change finishTrial, so that the on_finish callback is called before data.write and given the data object by reference, so that this trial-level callback can modify the data before it is saved.

This makes to me. It will simplify the process of modifying the trial data. It's a little more dangerous, in that it becomes easier to accidentally change the data, but I think the benefits outweigh this concern.

**Comment by jodeleeuw** [comment]
I think this is now addressed with the new DataColletion object.

The only unresolved piece is the timing of the on_data_update callback, which still occurs before the on_finish and on_trial_finish event handlers. I'm going to open a new issue to deal with that.
